mediatek: upstream-ready spi-nand driver

981213 opened the pull request:
Replace the mediatek spi-nand driver with a backport of spi-mem ECC framework and a upstream-ready spi-mem driver.

The patches themselves are submitted for review:
https://patchwork.kernel.org/project/spi-devel-general/list/?series=628772

(only doc related comments in v3 right now.)
Meanwhile I decided to open a pull-request here for more eyes on it.

This breaks 5.10 support for mediatek.
I think I'll switch mediatek to 5.15 when I'm ready to merge this.

981213 committed:
mediatek: drop spi-nand driver from mtk

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
mediatek: v5.15: backport spi-mem ecc support

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
mediatek: spi-nand: check for 1 byte BBM only

the OOB layout in MTK SNFI uses the 2nd byte, and anything using OOB
will make the block a bad-block in spi-nand driver.
Hack it for now. We need a proper solution upstream.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
mediatek: add an upstream-ready spi-nand driver

This patch implements the spi-nand controller driver as an ECC-capable
spi-mem controller to use the upstream SPI-NAND driver.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

dangowrt commented:
To avoid problems like last time when the SPI-NAND driver was swapped for what we have now, I thought I'd rather test it early this time ;)
Result: flash isn't detected at all.
```
[    0.562531] mtk-ecc 1100e000.ecc: probed
[    0.568529] mtk-snand 1100d000.spi: page format: (512 + 16) * 1
[    0.574742] spi-nand spi2.0: unknown raw ID e571e571
[    0.579744] spi-nand: probe of spi2.0 failed with error -524
```

This is because you removed the driver for Fidelix SPI-NAND by commit c6ed31630d6c9a1cb8b83a37bac77e480bc127b7
So for this to work at all, first I had to re-add it, this time for Linux 5.15.
[Fidelix-patch.tar.gz](https://github.com/openwrt/openwrt/files/8445320/Fidelix-patch.tar.gz)

```
[    0.563438] mtk-ecc 1100e000.ecc: probed
[    0.569437] mtk-snand 1100d000.spi: page format: (512 + 16) * 1
[    0.575657] spi-nand spi2.0: Fidelix SPI NAND was found.
[    0.581000] spi-nand spi2.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
[    0.589155] mtk-snand 1100d000.spi: page format: (512 + 16) * 4
[    0.595075] mtk-snand 1100d000.spi: ECC strength: 4 bits per 512 bytes
[    0.601901] 4 fixed-partitions partitions found on MTD device spi2.0
[    0.608268] Creating 4 MTD partitions on "spi2.0":
[    0.613051] 0x000000000000-0x000000080000 : "bl2"
[    0.618545] mtdblock: MTD device 'bl2' is NAND, please consider using UBI block devices instead.
[    0.627493] 0x000000080000-0x0000001c0000 : "fip"
[    0.633705] mtdblock: MTD device 'fip' is NAND, please consider using UBI block devices instead.
[    0.642648] 0x0000001c0000-0x0000002c0000 : "factory"
[    0.648979] mtdblock: MTD device 'factory' is NAND, please consider using UBI block devices instead.
[    0.658263] 0x000000300000-0x000008000000 : "ubi"
[    0.792414] mtdblock: MTD device 'ubi' is NAND, please consider using UBI block devices instead.
...
[    1.959668] UBI: auto-attach mtd3
[    1.962997] ubi0: attaching mtd3
[    2.236100] ubi0: scanning is finished
[    2.244247] ubi0: attached mtd3 (name "ubi", size 125 MiB)
[    2.249763] ubi0: PEB size: 131072 bytes (128 KiB), LEB size: 126976 bytes
[    2.256641] ubi0: min./max. I/O unit sizes: 2048/2048, sub-page size 2048
[    2.263421] ubi0: VID header offset: 2048 (aligned 2048), data offset: 4096
[    2.270378] ubi0: good PEBs: 1000, bad PEBs: 0, corrupted PEBs: 0
[    2.276463] ubi0: user volume: 6, internal volumes: 1, max. volumes count: 128
[    2.283679] ubi0: max/mean erase counter: 13/7, WL threshold: 4096, image sequence number: 1841606655
[    2.292893] ubi0: available PEBs: 482, total reserved PEBs: 518, PEBs reserved for bad PEB handling: 20
[    2.302351] ubi0: background thread "ubi_bgt0d" started, PID 381
[    2.309466] FIT: Selected configuration: "config-1" (OpenWrt linksys_e8450-ubi)
[    2.316847] FIT:           kernel sub-image 0x00001000..0x00520a37 "kernel-1" (ARM64 OpenWrt Linux-5.15.31) 
[    2.326688] FIT:          flat_dt sub-image 0x00521000..0x00528620 "fdt-1" (ARM64 OpenWrt linksys_e8450-ubi device tree blob) 
[    2.338089] FIT:       filesystem sub-image 0x00529000..0x00996fff "rootfs-1" (ARM64 OpenWrt linksys_e8450-ubi rootfs) 
[    2.348874] FIT: selecting configured loadable "rootfs-1" to be root filesystem
[    2.356173]  ubiblock0_4: p1(rootfs-1)
[    2.356367] block ubiblock0_4: created from ubi0:4(fit)
[    2.368851] VFS: Mounted root (squashfs filesystem) readonly on device 259:0.
...
```

Looks like it's all working. Also nvmem MAC addresses show up correctly and `mt76` correctly reads calibration data from NAND.

The only oddity I can spot is
```
[    0.569437] mtk-snand 1100d000.spi: page format: (512 + 16) * 1
```
before the SPI chip is even probed. It is then changed to
```
[    0.589155] mtk-snand 1100d000.spi: page format: (512 + 16) * 4
```
once the chip is been detected. I assume this is just left-over debugging output right?

981213 commented:
> This is because you removed the driver for Fidelix SPI-NAND by commit [c6ed316](https://github.com/openwrt/openwrt/commit/c6ed31630d6c9a1cb8b83a37bac77e480bc127b7) So for this to work at all, first I had to re-add it, this time for Linux 5.15. [Fidelix-patch.tar.gz](https://github.com/openwrt/openwrt/files/8445320/Fidelix-patch.tar.gz)

Oops. I forgot about this one.

> Looks like it's all working. Also nvmem MAC addresses show up correctly and `mt76` correctly reads calibration data from NAND.
> 
> The only oddity I can spot is
> 
> ```
> [    0.569437] mtk-snand 1100d000.spi: page format: (512 + 16) * 1
> ```
> 
> before the SPI chip is even probed. It is then changed to
> 
> ```
> [    0.589155] mtk-snand 1100d000.spi: page format: (512 + 16) * 4
> ```
> 
> once the chip is been detected. I assume this is just left-over debugging output right?

That's how this controller works: It must have a valid page format configured before it can perform certain ops.
It's first set up with the minimum supported page+spare size during probing and the actual format used is set when spi-nand calls ecc engine init.

981213 commented:
> I assume this is just left-over debugging output right?

I'll change this to dev_dbg to hide this detail if there's a need for a v5 upstream :)

981213 committed:
mediatek: enable snfi for mt7629

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

981213 committed:
mediatek: restore support for Fidelix FM35X1GA

This patch was dropped when the spi-nand driver got replaced last time.
Note: This isn't a complete chip support. It's missing ECC status
functions. It's fine on MT7622 because we don't use it, but this means
it can't be upstreamed.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

namidairo commented:
A couple issues with chip support in the old driver being a bit wider compared to upstream?

As far as I can tell, no ESMT F50L1G41LB (0xC8, 0x01) defined upstream, so this may break a fair few (Half? It's one of the two chips I've seen) Xiaomi AX6S units. The other half are Gigadevice GD5F1GQ5xExxG, which is supported in 5.15, so that's fine. (Would require a backport to 5.10 though)

A quick search shows it isn't too complex to add support for that particular ESMT one (https://github.com/rockchip-linux/u-boot/commit/52b0060178285488737854a48ddecd381f8b236e).

The "Ruijie RG-EW3200GX PRO" will probably have a similar issue, based on the device info.

981213 commented:
> A couple issues with chip support in the old driver being a bit wider compared to upstream?
> 
> As far as I can tell, no ESMT F50L1G41LB (0xC8, 0x01) defined upstream, so this may break a fair few (Half? It's one of the two chips I've seen) Xiaomi AX6S units.

Patch sent upstream:
https://patchwork.ozlabs.org/project/linux-mtd/patch/20220413042337.160263-1-gch981213@gmail.com/
I'll backport it when it's accepted.

> 
> A quick search shows it isn't too complex to add support for that particular ESMT one ([rockchip-linux/u-boot@52b0060](https://github.com/rockchip-linux/u-boot/commit/52b0060178285488737854a48ddecd381f8b236e)).
> 
> The "Ruijie RG-EW3200GX PRO" will probably have a similar issue, based on the device info.

This one uses SPI-NOR so it's not a problem.

namidairo commented:
> This one uses SPI-NOR so it's not a problem.

Oops, eyes went straight for the vendor names without even seeing it was only 16mb.

That said, I think the Elecom WRC-2533gent might have to be runtime tested too, since I can't find any boot log, board images etc, but it has spi-nand specified in the dts. It's probably an already supported Macronix or Winbond, though.

981213 committed:
mediatek: switch to 5.15

the SPI-NAND driver switch breaks dts compatibility. It's too much work
to backport all ECC framework support to 5.10 so let's switch the target
to 5.15 instead.

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>

dangowrt commented on target/linux/mediatek/image/mt7622.mk:
```diff
@@ -203,6 +203,16 @@ define Device/mediatek_mt7622-rfb1-ubi
+  DEVICE_DTS := mt7622-rfb-mtk0897-ubi
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := swconfig
+  IMAGE/sysupgrade.bin := sysupgrade-tar | append-metadata
```
As you have enabled `CONFIG_FIT_PARTITION` you could use a single uImage.FIT instead of sysupgrade.tar. See BPi-R64 and E8450/RT3200 for examples.

981213 replied:
I know about that one and I applied it to the mt7629 board.
I flashed this MT7622 reference board I have with the old u-boot-2014 and enabled ubi support in this ancient u-boot. (UBI on top of MTK BMT, just for testing.) The new FIT image doesn't work with the old u-boot.
BTW Neither device support works with the original bootloaders and I don't plan to push the two device support patches to master. That's why they don't have proper commit messages. Sorry for wasted reviewing efforts :D

dangowrt replied:
Yes, you need U-Boot 2018.01 or later for external-data FIT.

dangowrt commented:
Less promising results with 5.18.0-rc3-next-20220422 and you patches applied on BPi-R64:
```
[    0.813441] mtk-ecc 1100e000.ecc: probed
[    0.819296] spi-nand spi1.0: Winbond SPI NAND was found.
[    0.824671] spi-nand spi1.0: 128 MiB, block size: 128 KiB, page size: 2048, OOB size: 64
[    0.832828] mtk-snand 1100d000.spi: ECC strength: 4 bits per 512 bytes
[    0.841054] 3 fixed-partitions partitions found on MTD device spi1.0
[    0.847448] Creating 3 MTD partitions on "spi1.0":
[    0.852243] 0x000000000000-0x000000080000 : "bl2"
[    1.912175] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[    2.952152] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[    3.992156] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[    5.032147] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[    5.039179] mtdblock: MTD device 'bl2' is NAND, please consider using UBI block devices instead.
[    5.048142] 0x000000080000-0x000000280000 : "fip"
[    6.072178] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[    7.112152] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[    8.152145] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[    9.192154] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   10.232145] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   11.272150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   12.312144] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   13.352150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   14.392144] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   15.432150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   16.472145] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   17.512150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   18.552145] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   19.592150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   20.632144] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   21.672150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   21.679161] mtdblock: MTD device 'fip' is NAND, please consider using UBI block devices instead.
[   21.688284] 0x000000280000-0x000008000000 : "ubi"
[   22.712150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   23.752163] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   24.792145] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   25.832150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   26.872145] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   27.912150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   28.952144] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   29.992150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   31.032145] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   32.072150] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
[   33.112144] mtk-snand 1100d000.spi: DMA timed out for reading from cache.
...
```

This behavior could also be caused by my hack to launch partition parsers on plain mtdblock devices (ie. to have uImage.FIT partitions on devices with SPI-NOR like the UniFi 6 LR). However, it works fine with the old driver and also with your new driver on the E8450 with Fidelix SPI-NAND. Any ideas?

981213 commented:
@dangowrt 
try these two patches:
```diff
From 63e1ca8900e4ef78e9cafa9fadf5323a59057b6c Mon Sep 17 00:00:00 2001
From: Chuanhong Guo <gch981213@gmail.com>
Date: Sat, 23 Apr 2022 23:33:37 +0800
Subject: [PATCH] spi: mtk-snfi: add missing NFI mode switch

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>
---
 drivers/spi/spi-mtk-snfi.c | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/drivers/spi/spi-mtk-snfi.c b/drivers/spi/spi-mtk-snfi.c
index 7ba3c5a66bb6..2c556e304673 100644
--- a/drivers/spi/spi-mtk-snfi.c
+++ b/drivers/spi/spi-mtk-snfi.c
@@ -1398,6 +1398,9 @@ static int mtk_snand_probe(struct platform_device *pdev)
 		goto disable_clk;
 	}
 
+	// switch to SNFI mode
+	nfi_write32(ms, SNF_CFG, SPI_MODE);
+
 	// setup an initial page format for ops matching page_cache_op template
 	// before ECC is called.
 	ret = mtk_snand_setup_pagefmt(ms, ms->caps->sector_size,
-- 
2.35.1
```
```diff
From 616f15a4ff4f982ae9cce67dd552c54d10a5fc25 Mon Sep 17 00:00:00 2001
From: Chuanhong Guo <gch981213@gmail.com>
Date: Thu, 14 Apr 2022 20:01:57 +0800
Subject: [PATCH] spi: mtk-snfi: fix off-by-one in fifo filling

in mtk_snand_mac_io, after arranging all tx data, reg_offs is the total
tx length.
There's leftover data in val when the length isn't a multiple of 4,
not when (reg_offs % 4 != 3).

Signed-off-by: Chuanhong Guo <gch981213@gmail.com>
---
 drivers/spi/spi-mtk-snfi.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/spi/spi-mtk-snfi.c b/drivers/spi/spi-mtk-snfi.c
index 727057b6db89..7ba3c5a66bb6 100644
--- a/drivers/spi/spi-mtk-snfi.c
+++ b/drivers/spi/spi-mtk-snfi.c
@@ -480,7 +480,7 @@ static int mtk_snand_mac_io(struct mtk_snand *snf, const struct spi_mem_op *op)
 		}
 	}
 
-	if (reg_offs % 4 != 3)
+	if (reg_offs % 4)
 		nfi_write32(snf, SNF_GPRAM + (reg_offs & ~3), val);
 
 	for (i = 0; i < reg_offs; i += 4)
-- 
2.35.1
```

dangowrt commented:
> try these two patches:

That solves it. Now linux-next is still crashing because of recent changes in drivers/net/dsa/mt7530.c, but that's another story...
